drivers/video/video_arducam_mega.c

```diff
@@ -0,0 +1,1186 @@
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(640, 480, VIDEO_PIX_FMT_RGB565),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(1280, 720, VIDEO_PIX_FMT_RGB565),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(1600, 1200, VIDEO_PIX_FMT_RGB565),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(1920, 1080, VIDEO_PIX_FMT_RGB565),
+	{0},
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(96, 96, VIDEO_PIX_FMT_JPEG),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(128, 128, VIDEO_PIX_FMT_JPEG),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(320, 240, VIDEO_PIX_FMT_JPEG),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(320, 320, VIDEO_PIX_FMT_JPEG),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(640, 480, VIDEO_PIX_FMT_JPEG),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(1280, 720, VIDEO_PIX_FMT_JPEG),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(1600, 1200, VIDEO_PIX_FMT_JPEG),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(1920, 1080, VIDEO_PIX_FMT_JPEG),
+	{0},
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(96, 96, VIDEO_PIX_FMT_YUYV),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(128, 128, VIDEO_PIX_FMT_YUYV),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(320, 240, VIDEO_PIX_FMT_YUYV),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(320, 320, VIDEO_PIX_FMT_YUYV),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(640, 480, VIDEO_PIX_FMT_YUYV),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(1280, 720, VIDEO_PIX_FMT_YUYV),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(1600, 1200, VIDEO_PIX_FMT_YUYV),
+	ARDUCAM_MEGA_VIDEO_FORMAT_CAP(1920, 1080, VIDEO_PIX_FMT_YUYV),
+	{0},
+	{0},
+};
+
+#define SUPPORT_RESOLUTION_NUM 9
+
+static uint8_t support_resolution[SUPPORT_RESOLUTION_NUM] = {
+	MEGA_RESOLUTION_96X96,   MEGA_RESOLUTION_128X128, MEGA_RESOLUTION_QVGA,
+	MEGA_RESOLUTION_320X320, MEGA_RESOLUTION_VGA,     MEGA_RESOLUTION_HD,
+	MEGA_RESOLUTION_UXGA,    MEGA_RESOLUTION_FHD,     MEGA_RESOLUTION_NONE,
+};
+
+static int arducam_mega_write_reg(const struct spi_dt_spec *spec, uint8_t reg_addr, uint8_t value)
+{
+	uint8_t tries = 3;
+
+	reg_addr |= 0x80;
+
+	struct spi_buf tx_buf[2] = {
+		{.buf = &reg_addr, .len = 1},
+		{.buf = &value, .len = 1},
+	};
+
+	struct spi_buf_set tx_bufs = {.buffers = tx_buf, .count = 2};
+
+	while (tries--) {
+		if (!spi_write_dt(spec, &tx_bufs)) {
+			return 0;
+		}
+		/* If writing failed wait 5ms before next attempt */
+		k_msleep(5);
+	}
+	LOG_ERR("failed to write 0x%x to 0x%x", value, reg_addr);
+
+	return -1;
+}
+
+static int arducam_mega_read_reg(const struct spi_dt_spec *spec, uint8_t reg_addr)
+{
+	uint8_t tries = 3;
+	uint8_t value;
+	uint8_t ret;
+
+	reg_addr &= 0x7F;
+
+	struct spi_buf tx_buf[] = {
+		{.buf = &reg_addr, .len = 1},
+	};
+
+	struct spi_buf_set tx_bufs = {.buffers = tx_buf, .count = 1};
+
+	struct spi_buf rx_buf[] = {
+		{.buf = &value, .len = 1},
+		{.buf = &value, .len = 1},
+		{.buf = &value, .len = 1},
+	};
+
+	struct spi_buf_set rx_bufs = {.buffers = rx_buf, .count = 3};
+
+	while (tries--) {
+		ret = spi_transceive_dt(spec, &tx_bufs, &rx_bufs);
+		if (!ret) {
+			return value;
+		}
+		/* If reading failed wait 5ms before next attempt */
+		k_msleep(5);
+	}
+	LOG_ERR("failed to read 0x%x register", reg_addr);
+
+	return -1;
```

Comment:
This needs to be returning the original `ret` to propagate the error, helpful for debugging or ignoring selectively some errors (i.e. retry on `-EAGAIN`, or `-EIO` from the application).